pkg/controllers/management/eks/eks_upstream_refresh.go

```diff
@@ -0,0 +1,178 @@
+package eks
+
+import (
+	"reflect"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/service/eks"
+	"github.com/rancher/eks-operator/controller"
+	v1 "github.com/rancher/eks-operator/pkg/apis/eks.cattle.io/v1"
+	v3 "github.com/rancher/rancher/pkg/generated/controllers/management.cattle.io/v3"
+	mgmtv3 "github.com/rancher/rancher/pkg/generated/norman/management.cattle.io/v3"
+	"github.com/rancher/rancher/pkg/wrangler"
+	wranglerv1 "github.com/rancher/wrangler/pkg/generated/controllers/core/v1"
+	"github.com/robfig/cron"
+	"github.com/sirupsen/logrus"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/runtime"
+)
+
+type eksRefreshController struct {
+	refreshCronJob *cron.Cron
+	secretsCache   wranglerv1.SecretCache
+	clusterClient  v3.ClusterClient
+	clusterCache   v3.ClusterCache
+}
+
+func RegisterRefresh(wContext *wrangler.Context) {
+	e := &eksRefreshController{
+		refreshCronJob: cron.New(),
+		secretsCache:   wContext.Core.Secret().Cache(),
+		clusterClient:  wContext.Mgmt.Cluster(),
+		clusterCache:   wContext.Mgmt.Cluster().Cache(),
+	}
+
+	schedule, err := cron.ParseStandard("*/5 * * * *")
+	if err != nil {
+		logrus.Errorf("Error parsing auth refresh cron: %v", err)
+		return
+	}
+	e.refreshCronJob.Schedule(schedule, cron.FuncJob(e.refreshAllUpstreamStates))
+	e.refreshCronJob.Start()
+}
+
+func (e *eksRefreshController) refreshAllUpstreamStates() {
+	logrus.Infof("Refreshing EKS clusters' upstream states")
+	clusters, err := e.clusterCache.List(labels.Everything())
+	if err != nil {
+		logrus.Error("error trying to refresh EKS clusters' upstream states")
+	}
+
+	for _, cluster := range clusters {
+		if _, err := e.refreshClusterUpstateSpec(cluster); err != nil {
+			logrus.Errorf("error refreshing EKS cluster [%s] upstream state", cluster.Name)
+		}
+	}
+}
+
+func (e *eksRefreshController) refreshClusterUpstateSpec(cluster *mgmtv3.Cluster) (*mgmtv3.Cluster, error) {
+	if cluster == nil || cluster.DeletionTimestamp != nil {
+		return nil, nil
+	}
+
+	if cluster.Spec.EKSConfig == nil {
+		return cluster, nil
+	}
+
+	logrus.Infof("checking cluster [%s] upstream state for changes", cluster.Name)
+
+	if cluster.Status.EKSStatus.UpstreamSpec == nil {
+		logrus.Infof("initial upstream spec for cluster [%s] has not been set by eks cluster handler yet, skipping", cluster.Name)
+		return cluster, nil
+	}
+
+	upstreamSpec, err := getComparableUpstreamSpec(e.secretsCache, cluster)
+	if err != nil {
+		return cluster, err
+	}
+
+	if !reflect.DeepEqual(cluster.Status.EKSStatus.UpstreamSpec, upstreamSpec) {
+		logrus.Infof("updating cluster [%s], upstream change detected", cluster.Name)
+		cluster = cluster.DeepCopy()
+		cluster.Status.EKSStatus.UpstreamSpec = upstreamSpec
+		cluster, err = e.clusterClient.Update(cluster)
+		if err != nil {
+			return cluster, err
+		}
+	}
+
+	if !reflect.DeepEqual(cluster.Spec.EKSConfig, cluster.Status.AppliedSpec.EKSConfig) {
```

Comment:
Would it make sense to do this check first before attempting to pull the EKS config from AWS?

Reply:
No, because we still want to write the upstream state to the status we just do not want to overwrite spec.